README.md
<!--
 * @Author: gaoyong06@qq.com 
 * @Date: 2020-11-30 12:56:02 
 * @Last Modified by: gaoyong06@qq.com
 * @Last Modified time: 2020-11-30 12:56:02
-->
|      |      |
| ---- | ---- |
|  作者    |  [gaoyong](mailto:gaoyong06@qq.com) |
|  创建时间    | 2020-11-30 12:56:02      |

# data_finder

字节跳动[数据漫游者](https://datarangers.com.cn/)的基础分析产品[数据发现者DataFinder](https://datarangers.com.cn/product/datafinder)的Flutter插件

## 介绍

data_finder是一个flutter plugin，为满足在使用flutter开发的App上使用字节跳动提供的[数据漫游者](https://datarangers.com.cn/)的基础分析产品[数据发现者DataFinder](https://datarangers.com.cn/product/datafinder)而开发的一个flutter plugin，目前支持Android, iOS两个平台
## 主要功能

- 设备注册/激活
  - 初始化dataFinder
  - 用户登录后设置登录用户uuid
- 手动上报用户行为日志
  - 上报行为埋点

## 安装
- 在项目的pubspec.yaml中在增加下面配置：data_finder: ^0.0.7
- 然后在项目根目录(pubspec.yaml 所在的目录)执行命令：flutter pub get

## 使用方法
 - 设备注册/激活
```
  
  //初始化SDK
  //备注：204291为申请的appId
  DataFinder.start("204291", channel: "App Store", showLog: true)

  //登录态变化调用(设置uuid)
  //备注：sf180为系统自身的uid
  DataFinder.setUserUniqueID("sf180")

```
  - 手动上报用户行为日志
```
  
   DataFinder.onEventV3("event_name", params: {
      "params1": "value1",
      "params2": "value2",
   }

```  

**初始化dataFinder** 

方法名称: DataFinder.start

参数说明:

| 参数名称      | 说明                         | 类型               | 默认值 |
| ------------- | ---------------------------- | ------------------ | ------ |
| appId          | 在dataFinder后台申请的appId(应用列表中的应用ID) | String            | -      |
| channel        | 渠道名称,iOS一般默认App Store                   | String             | app     |
| showLog         | 是否在控制台输出日志，可用于观察用户行为日志上报情况          | bool | false     |

示例：
```
  //初始化SDK
  //备注：204291为申请的appId
  DataFinder.start("204291", channel: "App Store", showLog: true)
```

**用户登录后设置登录用户uuid** 

方法名称: DataFinder.setUserUniqueID

参数说明: 

| 参数名称      | 说明                         | 类型               | 默认值 |
| ------------- | ---------------------------- | ------------------ | ------ |
| userUniqueID          | 自己的账号体系ID, 并保证其唯一性 | String            | -      |

示例：
```
  //登录态变化调用(设置uuid)
  //备注：sf180为系统自身的uid
  DataFinder.setUserUniqueID("sf180")
```

**上报行为埋点** 

方法名称: DataFinder.onEventV3

参数说明: 

| 参数名称      | 说明                         | 类型               | 默认值 |
| ------------- | ---------------------------- | ------------------ | ------ |
| event          | 事件名称 | String            | -      |
| params        | 事件属性,一个事件可以对应多个属性                   | Map             |      |

示例：
```
  DataFinder.onEventV3("event_name", params: {
      "params1": "value1",
      "params2": "value2",
   }
```
[示例程序](https://github.com/gaoyong06/data_finder/blob/main/example/lib/main.dart)

## iOS平台配置

- 修改Podfile文件
  
  在/Users/gaoyong/Documents/work/data_finder/example/ios/Podfile 文件中增加:

  ```
  source 'https://github.com/bytedance/cocoapods_sdk_source_repo.git'
  ```

  因为RangersAppLog5.6.3 现在(2020/12/08)还有bug,这里强制使用5.6.1
  ```
  pod 'RangersAppLog','5.6.1'
  ```

  ```
  pre_install do |installer|
    # workaround for https://github.com/CocoaPods/CocoaPods/issues/3289
    Pod::Installer::Xcode::TargetValidator.send(:define_method, :verify_no_static_framework_transitive_dependencies) {}
  end
  ```

  [示例程序](https://github.com/gaoyong06/data_finder/blob/main/example/ios/Podfile)

- 执行pod install --repo-update
  
  在/Users/gaoyong/Documents/work/data_finder/example/ 目录下执行
  
  ```
  pod install --repo-update
  ```

## Android平台配置

- 无需配置
  
## 配置的其他问题

- 理论上按上面的配置就可以正常工作了，如果出现其他问题可能会因为版本或者环境问题导致，可先阅读一下DataFinder的提供的应用[接入文档](https://datarangers.com.cn/help/doc?lid=1097&did=8547)

- 因为现在没有涉及URL Scheme，所以上面没有配置，如果有需要参考接入文档,[iOS添加URL Scheme](https://datarangers.com.cn/help/doc?lid=1097&did=8547#_4-%E6%B7%BB%E5%8A%A0url-scheme) 和[Android添加url scheme](https://datarangers.com.cn/help/doc?lid=1097&did=10942#_2-3-%E6%B7%BB%E5%8A%A0url-scheme)

- 如果在接入过程中遇到其他问题欢迎大家提交issue,或者pr

## 执行示例程序

  如果希望执行一下该项目的示例程序的话看看效果的话，clone 或者 fork 该项目到本地硬盘后，cd 到 ../data_finder/example 目录下后，如果想看一下Android的效果，则cd至android 目录（../data_finder/example/android）执行 flutter run, 如果想看一下iOS的效果，用xcode 打开../data_finder/example/ios/Runner.xcworkspace, 编译执行即可

## 在ios上测试使用的命令

- flutter build ios --no-codesign
- flutter run

## 开发过程中遇到的问题及处理办法

- [!] CocoaPods did not set the base configuration of your project because your project already has a custom config set. In order for CocoaPods integration to work at all, please either set the base configurations of the target `Runner` to `Target Support Files/Pods-Runner/Pods-Runner.profile.xcconfig` or include the `Target Support Files/Pods-Runner/Pods-Runner.profile.xcconfig` in your build configuration (`Flutter/Release.xcconfig`).
  
  >处理办法：[参考文档](https://github.com/flutter/flutter/issues/49495#issuecomment-583066933)

- [!] The 'Pods-Runner' target has transitive dependencies that include statically linked binaries: (RangersAppLog)
  
  >处理办法：修改/Users/gaoyong/Documents/work/data_finder/example/ios/Podfile 注释掉 use_frameworks!
  将use_frameworks! 修改为 #use_frameworks!

- [!] Automatically assigning platform `iOS` with version `9.0` on target `Runner` because no platform was specified. Please specify a platform for this target in your Podfile. See `https://guides.cocoapods.org/syntax/podfile.html#platform`.

  >处理办法：修改/Users/gaoyong/Documents/work/data_finder/example/ios/Podfile 打开注释 # platform :ios, '9.0'
  将# platform :ios, '9.0' 修改为 platform :ios, '9.0'

- Command PhaseScriptExecution failed with a nonzero exit code
  > 处理办法：Runner project > Info > Configurations 设置错误，正确的设置方法，参考这个[设置和验证方式](https://github.com/flutter/flutter/issues/49495#issuecomment-583066933)

- Flutter/Debug.xcconfig:2: could not find included file 'Generated.xcconfig' in search paths

  > 处理办法：https://stackoverflow.com/questions/54321180/error-could-not-find-included-file-generated-xcconfig-in-search-paths-in-tar
  另：如果没有设置签名，则执行 flutter build ios --no-codesign

## 备注
    - *上面文档中提到的目录/Users/gaoyong/Documents/work/data_finder/example 仅为示例，实际场景应该是开发者自己的项目根目录(即含有pubspec.yaml文件的目录)*  

CHANGELOG.md
<!--
 * @Author: gaoyong06@qq.com 
 * @Date: 2020-11-30 12:56:02 
 * @Last Modified by: gaoyong06@qq.com
 * @Last Modified time: 2020-11-30 12:56:02
-->
## 0.0.7

  * Update documentation
## 0.0.6

  * Bugfix：iOS Compile fail
## 0.0.5

  * Update documentation
## 0.0.4

  * Update documentation
## 0.0.3

  * Update documentation
  * Bugfix：Android Compile fail

## 0.0.2

  * Add support for ios platform
## 0.0.1

  * Initial Version of the library

test/data_finder_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:data_finder/data_finder.dart';

void main() {
  const MethodChannel channel = MethodChannel('data_finder');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await DataFinder.platformVersion, '42');
  });
}


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:data_finder_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


example/lib/main.dart
/*
 * @Author: gaoyong06@qq.com 
 * @Date: 2020-11-30 12:56:02 
 * @Last Modified by: gaoyong06@qq.com
 * @Last Modified time: 2020-11-30 12:56:02
 */
import 'package:flutter/material.dart';
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:data_finder/data_finder.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _platformVersion = 'Unknown';

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  // Platform messages are asynchronous, so we initialize in an async method.
  Future<void> initPlatformState() async {
    String platformVersion;
    // Platform messages may fail, so we use a try/catch PlatformException.
    try {
      platformVersion = await DataFinder.platformVersion;
    } on PlatformException {
      platformVersion = 'Failed to get platform version.';
    }

    // If the widget was removed from the tree while the asynchronous platform
    // message was in flight, we want to discard the reply rather than calling
    // setState to update our non-existent appearance.
    if (!mounted) return;

    setState(() {
      _platformVersion = platformVersion;
    });
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Column(
            children: [
              Text('Running on: $_platformVersion\n'),
              RaisedButton(
                  onPressed: () => DataFinder.start("204291", showLog: true),
                  child: Text("start")),
              RaisedButton(
                  onPressed: () => DataFinder.setUserUniqueID("sf180"),
                  child: Text("setUserUniqueID")),
              RaisedButton(
                onPressed: () =>
                    DataFinder.onEventV3("predefine_pageview", params: {
                  "url_path": "article/article_detail",
                  "title": "页面浏览",
                }),
                child: Text("onEventV3-Page"),
              ),
              RaisedButton(
                onPressed: () => DataFinder.onEventV3("click_button", params: {
                  "name": "btn_click",
                  "label": "按钮点击",
                }),
                child: Text("onEventV3-Btn"),
              ),
            ],
          ),
        ),
      ),
    );
  }
}


ios/Classes/SwiftDataFinderPlugin.swift
/*
 * @Author: gaoyong06@qq.com
 * @Date: 2020-12-02 18:27:23
 * @Last Modified by: gaoyong06@qq.com
 * @Last Modified time: 2020-12-02 18:27:23
 */
import Flutter
import UIKit
import RangersAppLog

//datafinder ios插件类
//参考文档：https://datarangers.com.cn/help/doc?lid=1097&did=8547
public class SwiftDataFinderPlugin: NSObject, FlutterPlugin {
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "data_finder", binaryMessenger: registrar.messenger())
    let instance = SwiftDataFinderPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    
    switch call.method {
    case "getPlatformVersion":
        result("iOS " + UIDevice.current.systemVersion);
    case "start":
        start(call: call, result: result);
    case "setUserUniqueID":
        setUserUniqueID(call: call, result: result);
    case "onEventV3":
        onEventV3(call: call, result: result);
    default:
        result(FlutterMethodNotImplemented);
    }
  }
    
  //初始化SDK
  private func start(call: FlutterMethodCall, result: FlutterResult) {
    
    let arguments: [String:Any] = call.arguments as! [String:Any]
    print("start run, \(arguments)");
    let appId: String = arguments["appId"] as? String ?? "";
    let channel: String = arguments["channel"] as? String ?? ""
    let showLog: Bool = arguments["showLog"] as? Bool ?? false
    let appName: String = arguments["appName"] as? String ?? ""
    
    // Override point for customization after application launch.
    let config = BDAutoTrackConfig(appID: appId) // 如不清楚请联系专属客户成功经理
    config.appName = appName // 与您申请APPID时的app_name一致

    /* 只支持上报到国内: BDAutoTrackServiceVendorCN，上报国外请使用 BDAutoTrackServiceVendorSG，5.5.0 及以上版本可以参考第 2 节的说明*/
    config.serviceVendor = BDAutoTrackServiceVendor.CN;
    config.channel = channel // iOS一般默认App Store
    config.showDebugLog = showLog // 是否在控制台输出日志，仅调试使用。release版本请设置为 false
    config.logNeedEncrypt = true // 是否加密日志，默认加密。release版本请设置为 true
    config.autoTrackEnabled = true
    
    //TODO: 2020-12-02 17:50:11.685114+0800 Runner[23448:265034] [] nw_protocol_get_quic_image_block_invoke dlopen libquic failed
    BDAutoTrack.start(with: config)
    return result(true)
  }
    
  //登录态变化调用(设置uuid)
  private func setUserUniqueID(call: FlutterMethodCall, result: FlutterResult) {
    
    let arguments: [String:Any] = call.arguments as! [String:Any]
    print("setUserUniqueID run, \(arguments)");
    let userUniqueID: String = arguments["userUniqueID"] as? String ?? ""
    BDAutoTrack.setCurrentUserUniqueID(userUniqueID)
    return result(true)
  }
    
  //通过代码上报行为埋点
  private func onEventV3(call: FlutterMethodCall, result: FlutterResult) {
    
    let arguments: [String:Any] = call.arguments as! [String:Any]
    print("onEventV3 run, \(arguments)");
    let event: String = arguments["event"] as? String ?? ""
    let params:[String:Any]? = arguments["event"] as? [String:Any] ?? nil
    BDAutoTrack.eventV3(event, params: params)
    return result(true)
  }
}


android/src/main/kotlin/cn/com/datarangers/data_finder/DataFinderPlugin.kt
package cn.com.datarangers.data_finder

import android.content.Context
import android.text.TextUtils
import androidx.annotation.NonNull
import com.bytedance.applog.AppLog
import com.bytedance.applog.ILogger
import com.bytedance.applog.InitConfig
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.json.JSONException
import org.json.JSONObject
import kotlin.math.log


/** DataFinderPlugin */
class DataFinderPlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "data_finder")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
            "start" -> start(call, result)
            "setUserUniqueID" -> setUserUniqueID(call, result)
            "onEventV3" -> onEventV3(call, result)
            else -> result.notImplemented()
        }
    }


    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }


    private fun start(@NonNull call: MethodCall, @NonNull result: Result) {
        val appId = call.argument<String>("appId")
        if (TextUtils.isEmpty(appId)) {
            println("appId不能为空")
            result.error("-1", "appId不能为空", "请检查appId")
            return
        }
        val channel: String = call.argument<String>("channel") ?: ""
        val showLog: Boolean = call.argument<Boolean>("showLog") ?: false
        println("appId: $appId===========channel: $channel")
        val logger = ILogger { s, _ -> if (showLog) println(s) }
        AppLog.init(context, InitConfig(appId!!, channel).setAutoStart(true).setLogger(logger))
        result.success(true)
    }

    private fun setUserUniqueID(@NonNull call: MethodCall, @NonNull result: Result) {
        val userUniqueID = call.argument<String>("userUniqueID")
        if (TextUtils.isEmpty(userUniqueID)) {
            println("userUniqueID不能为空")
            result.error("-1", "userUniqueID不能为空", "请检查userUniqueID")
            return
        }
        println("userUniqueID: $userUniqueID===========")
        AppLog.setUserUniqueID(userUniqueID);
        result.success(true)
    }

    private fun onEventV3(@NonNull call: MethodCall, @NonNull result: Result) {
        val event = call.argument<String>("event")
        if (TextUtils.isEmpty(event)) {
            println("event不能为空")
            result.error("-1", "event不能为空", "请检查event")
            return
        }
        val params = call.argument<Map<String, *>>("params")
        println("event: $event===========params: $params")
        if (params == null) {
            AppLog.onEventV3(event!!)
        } else {
            val paramsObj = JSONObject()
            try {
                for ((key, value) in params) {
                    paramsObj.put(key, value)
                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
            AppLog.onEventV3(event!!, paramsObj)
        }
        result.success(true)
    }

}


lib/data_finder.dart
/*
 * @Author: gaoyong06@qq.com 
 * @Date: 2020-11-30 12:56:02 
 * @Last Modified by: gaoyong06@qq.com
 * @Last Modified time: 2020-11-30 12:56:02
 */
import 'dart:async';
import 'package:flutter/services.dart';

class DataFinder {
  static const MethodChannel _channel = const MethodChannel('data_finder');

  static Future<String> get platformVersion async {
    final String version = await _channel.invokeMethod('getPlatformVersion');
    return version;
  }

  ///
  /// 初始化dataFinder
  /// String appId: 在dataFinder后台申请的appId(应用列表中的应用ID)
  /// String channel: 渠道名称,iOS一般默认App Store
  /// bool showLog: 是否在控制台输出日志，可用于观察用户行为日志上报情况
  static void start(String appId,
      {String channel = 'app', bool showLog = false}) {
    _channel.invokeMethod(
        "start", {"appId": appId, "channel": channel, "showLog": showLog});
  }

  ///
  /// 用户登录后设置登录用户uuid
  /// String userUniqueID: 自己的账号体系ID, 并保证其唯一性
  static void setUserUniqueID(String userUniqueID) {
    _channel.invokeMethod("setUserUniqueID", {
      "userUniqueID": userUniqueID,
    });
  }

  ///
  /// 上报行为埋点
  /// String event: 事件名称
  /// Map params: 事件属性,一个事件可以对应多个属性
  static void onEventV3(String event, {Map params}) {
    _channel.invokeMethod("onEventV3", {
      "event": event,
      "params": params,
    });
  }
}
